src/gateway-user.ts

```diff
@@ -0,0 +1,358 @@
+  get dappchainGateway(): Contracts.LoomCoinTransferGateway {
+    return this._dappchainGateway
+  }
+
+  /**
+   * Deposits funds from mainnet to the gateway
+   */
+  async depositAsync(amount: BN): Promise<ethers.ContractTransaction> {
+    let currentApproval = await this._ethereumLoom.functions.allowance(
+      await this.ethAddress,
+      this._ethereumGateway.address
+    )
+
+    let currentApprovalBN = new BN(currentApproval.toString())
+
+    log('Current approval:', currentApproval)
+    if (amount.gt(currentApprovalBN)) {
+      let tx = await this._ethereumLoom.functions.approve(
+        this._ethereumGateway.address,
+        amount.sub(currentApprovalBN).toString()
+      )
+      await tx.wait()
+      log('Approved an extra', amount.sub(currentApprovalBN))
+    }
+    return this._ethereumGateway.functions.depositERC20(
+      amount.toString(),
+      this._ethereumLoom.address
+    )
+  }
+
+  /**
+   * Withdraw funds from the gateway to mainnet
+   */
+  async withdrawAsync(amount: BN): Promise<ethers.ContractTransaction> {
+    const sig = await this.depositCoinToDAppChainGatewayAsync(amount)
+    return this.withdrawCoinFromDAppChainGatewayAsync(amount, sig)
+  }
+
+  async resumeWithdrawalAsync() {
+    const receipt = await this.getPendingWithdrawalReceiptAsync()
+    if (receipt === null) {
+      log('No pending receipt')
+      return
+    }
+    const signature = CryptoUtils.bytesToHexAddr(receipt.oracleSignature)
+    const amount = receipt.tokenAmount!
+    return this.withdrawCoinFromDAppChainGatewayAsync(amount, signature)
+  }
+
+  async getPendingWithdrawalReceiptAsync(): Promise<IWithdrawalReceipt | null> {
+    return this._dappchainGateway.withdrawalReceiptAsync(this.loomAddress)
+  }
+
+  /**
+   * Retrieves the  DAppChain LoomCoin balance of a user
+   * @param address The address to check the balance of. If not provided, it will check the user's balance
+   */
+  async getDAppChainBalanceAsync(address: string | undefined): Promise<BN> {
+    // if no address is provided, return our balance
+    if (address === undefined) {
+      return this._dappchainLoom.getBalanceOfAsync(this.loomAddress)
+    }
+    const addr = this.prefixAddress(address)
+    const balance = await this._dappchainLoom.getBalanceOfAsync(addr)
+    return balance
+  }
+
+  /**
+   * Deposits an amount of LOOM tokens to the dappchain gateway and return a signature which can be used to withdraw the same amount from the mainnet gateway.
+   *
+   * @param amount The amount that will be deposited to the DAppChain Gateway (and will be possible to withdraw from the mainnet)
+   */
+  private async depositCoinToDAppChainGatewayAsync(amount: BN): Promise<string> {
+    let pendingReceipt = await this.getPendingWithdrawalReceiptAsync()
+    let signature: Uint8Array
+    if (pendingReceipt === null) {
+      await this._dappchainLoom.approveAsync(this._dappchainGateway.address, amount)
+      const ethereumAddressStr = await this.ethAddress
```

Comment:
Why is there an `await` here?

Reply:
Waiting for the approve promise to resolve, then move on